### DISCUSSION TITLE
Job running with old instance of code

### DISCUSSION THREAD

**Discussion by hello2pankaj** [quartznet/quartznet#2359] [Q&A]
I am using Quartz.Extensions.Hostin 3.8.1 scheduling package in .netcore6 web api application. App deployment environment is azure appservice (running on 2 instances). I have below quartz config for 4 jobs.  

```
public static void ConfigureQuartzJobs(this WebApplicationBuilder builder)
{
	var services = builder.Services;
	services.AddQuartzHostedService(opt => opt.WaitForJobsToComplete = true);
	services.AddQuartz(q =>
	{
		q.SchedulerId = "SomeId";
		q.SchedulerName = "SomeName";
		q.UsePersistentStore(s =>
				{
				  s.UseProperties = true;
				  s.UseClustering();
				  s.UseSqlServer(connectionString);
				  s.UseNewtonsoftJsonSerializer();
				});
				q.AddJob<job1>(j => j.StoreDurably().WithIdentity(jobkey1))
				 .AddTrigger(trigger => trigger
						   .ForJob(jobkey1)
						   .WithIdentity("tr1")
						   .WithSimpleSchedule(schedule => schedule.WithIntervalInMinutes(5).RepeatForever()));
						   

				q.AddJob<job2>(j => j.StoreDurably().WithIdentity(jobkey2))
				 .AddTrigger(trigger => trigger
						   .ForJob(jobkey3)
						   .WithIdentity("tr2")
						   .WithSimpleSchedule(schedule => schedule.WithIntervalInMinutes(5).RepeatForever()));
						   

				q.AddJob<job3>(j => j.StoreDurably().WithIdentity(jobkey3))
				 .AddTrigger(trigger => trigger
						   .ForJob(jobkey3)
						   .WithIdentity("tr3")
						   .WithSimpleSchedule(schedule => schedule.WithIntervalInMinutes(5).RepeatForever()));
						   

				q.AddJob<job4>(j => j.StoreDurably().WithIdentity(jobkey4))
				 .AddTrigger(trigger => trigger
						   .ForJob(jobkey4)
						   .WithIdentity("tr4")
						   .WithSimpleSchedule(schedule => schedule.WithIntervalInMinutes(60).RepeatForever()));
	});
}

```
Let say job1 calls instance of LogicClass which has method "Process(bool CalledByJob)". It is called by Job1 class something like below.

```
[DisallowConcurrentExecution]
public class Job1 : IJob
{
	public async Task Execute(IJobExecutionContext context)
	{
		try
		{
		   logicClass.Process(True);
		}
		catch (Exception ex)
		{
			await Console.Out.WriteLineAsync($"{context.FireTimeUtc}, {ex}. FireInstanceId={context.FireInstanceId}");
		}
	}
}

```

LogicClass's process method has code to log saying "Processing start. CalledByJob={calledbyJob}"

So when the job starts, i can see the log happening which is expected.

However , when i was done with testing, i removed the part "CalledByJob={calledbyJob}" from the log.  The log now only said "Processing start".

The issue i am seeing is that after deployment of the code (i.e without CalledByJob log), some job run is showing it (i.e. "calledByJob=true") in the log and some not. I tried restarting the azure app service and Purging the azure front door cache. no luck though. 

Can anyone please guide me why this is happening and what's the solution.

**Comment by hello2pankaj** [comment]
Hi @lahma , what should i do to resolve this? See below screen grab. I deployed the code to include the build number (BuildNumber=1.2405.23.2) in the log. then again deployed the code so that new build number (BuildNumber=1.2405.23.3) start appearing in the log. However, it is seen that the quartz.net is running code from both the builds. See below. Green one is latest build and yellow is the earlier build. I tried a lot since my original post above but this is always the case. Restarting server also doesn't help. Can you please guide.

Additional info - I am not using static variables. Job runs for not more than 1 second. Logs are in by time desc so i should see build 1.2405.23.3 as the top entry but seeing 1.2405.23.2.

![image](https://github.com/quartznet/quartznet/assets/15063036/397fcce7-7ec1-489e-a5e6-506773d3e5db)

**Reply by lahma** [reply]
Well Quartz doesn't handle the deployments or the running instances. It just picks the DLLs it has when it starts.

**Reply by hello2pankaj** [reply]
Got the issue. It was from my side. Our azure app services has slot. The code was running from both the slots and hence running in parallel. Thanks for replying though.

**Reply by lahma** [reply]
That would explain it. As said Quartz doesn't know much about the surrounding deployments.

**Reply by hello2pankaj** [reply]
Ok @lahma , i am seeing an issue. As i mentioned , I have enabled azure app service slot feature. the idea is any new build is first deployed in stage slot . tested there and then swapped with prod slot. This means that the new build number always goes to stage slot first and then to prod slot. Hence, i was seeing both the build numbers in job logs as i am logging them. Both the slots share same DB though.

To solve the issue , i added a environment variable which will tell me if job gets invoked from stage or prod slot. In ITriggerListener's VetoJobExecution implementation i am checking if its not prod slot, then i am vetoing the job.

Now i have this scenario. Prod slot is running **version2**.  I did changes to my job (configured for running every 5 mins) and deployed the build to stage slot for testing (not deployed to prod slot yet). Hence stage slot version is now **vesion3**. 

However i am seeing is job does run every 5 minutes, but either from stage slot or prod slot. Expectation is 

- it should run from prod slot  with **version2** , every 5 mins irrespective of new build in stage slot which is **version3** 
- When i swap the slots, prod slot should run now with **version3**

I am seeing in my logs like 

3.05 -- version 2
3.10 -- version 3
3.15 -- version 3
3.20 -- version 2
3.25 -- version 3

Hope i am able to explain the situation.

What should i do so that the prod slot runs continuously every 5 mins.

**Reply by lahma** [reply]
So maybe just don't start the scheduler if you are in the staging slot?

**Reply by hello2pankaj** [reply]
The prod slot need to run continuously as the job does important stuff which are time sensitive.

**Reply by lahma** [reply]
Yes,  run in production, not in staging.

**Reply by hello2pankaj** [reply]
Did not understood. Its quartz.net which is deciding to run from stage slot or prod slot. 

Also here, if i am not clear, when i say stage slot i am not talking about staging environment. All my environment (test, stage, prod) has 2 slots (stage and prod). So i am talking about slots and not  environment. hope i am making sense.

**Reply by lahma** [reply]
Quartz doesn't decide to start/run. You are starting it via code.

**Reply by hello2pankaj** [reply]
Understood . but how to resolve this scenario.

**Comment by daver77** [comment]
Hi @hello2pankaj, did you find a way to resolve this? I have the same issue, my AppService has 2 slots and my jobs run on both 'staging' and production slots.

**Reply by daver77** [reply]
Here's how I did it in case anyone finds this in the future!

Like @lahma says above, you need to prevent it running in the staging slot. To do so you can check the WEBSITE_DEFAULT_HOSTNAME appSetting, this is usually the internal domain name for the appService. In prod it will be something like yourwebsite.azurewebsites.net, in staging it will be yourwebsite-staging.azurewebsites.net, so we just need to check for staging in WEBSITE_DEFAULT_HOSTNAME.

```
// Prevent the job from starting on the staging slot
string hostname = config.GetValue<string>("WEBSITE_DEFAULT_HOSTNAME");
bool runJobs = hostname == null || !hostname.Contains("staging");

if (runJobs)
{
    services.AddQuartz(q =>
    {
    ......
    }
}
```

Depending how your release pipelines are setup you may need to deploy twice to get your code into the staging slot.
